### PR TITLE
[RFC] Improve errors for fields overriding destructor attributes

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
Compiler generated destructors calling both field and user-defined
destructors can have less-restrictive attributes than the user-defined
one. This caused confusing error messages stating that the destructor
is not `pure`albeit the user marked their implementation as `pure`.

This commit adds an additional error message explaining the situation
and also lists the fields causing the attribute missmatch s.t. the user
may fix their respective destructors.

---

TL;DR: The compiler tells you why your code doesn't compile even though your non-rewritten code should - especially important for `-preview=fielddtors`.

### CODE REVIEW

**Comment by kinke** [comment]
Completely unrelated, but with the introduction of `frontend.h`, every new `extern(D)` function addition requiring a change and me being a code-owner of *.h, I will now probably be requested for review for almost every PR. Can we get rid of these `// ignoring extern () block because of linkage` comments (which don't even mention the name of the function and are so IMO hardly useful anyway)?

**Comment by MoonlightSentinel** [comment]
Yeah, that's annoying to update as well.

File: src/dmd/expression.d [hunk 1]
```diff
@@ -1107,11 +1107,81 @@ extern (C++) abstract class Expression : ASTNode
...
+     * which passes `check` while `f` doesn't (e.g. when the user defined dtor is pure but
+     * the generated dtor is not).
+     * In that case the method will identify and print all members causing the attribute
+     * missmatch.
+     *
+     * Params:
+     *   sc = scope
+     *   f  = potential `DtorDeclaration`
+     *   check = current check (e.g. whether it's pure)
+     *   what = the kind of check (e.g. `"pure"`)
```
Comment (wilzbach) [inline src/dmd/expression.d:1128]:
```suggestion
     *   checkName = the kind of check (e.g. `"pure"`)
```

File: test/fail_compilation/dtor_attributes.d [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+fail_compilation\dtor_attributes.d(410): Error: `pure` function `dtor_attributes.test4` cannot call impure destructor `dtor_attributes.StrictNested.~this`
+fail_compilation\dtor_attributes.d(410):        it's impure because of the following field's destructors:
+fail_compilation\dtor_attributes.d(410):         - arrayMember
+---
+*/
+#line 400
+
+struct StrictNested
+{
+    HasDtor[4] arrayMember;
```
Comment (wilzbach) [inline test/fail_compilation/dtor_attributes.d:116]: Could you also add `HasDtorClass[4] classMember;` (for completeness)

File: test/fail_compilation/dtorfields_attributes.d [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+fail_compilation\dtorfields_attributes.d(117): Error: `@safe` constructor `dtorfields_attributes.Strict.this` cannot call `@system` destructor `dtorfields_attributes.Strict.~this`
+fail_compilation\dtorfields_attributes.d(119):        `dtorfields_attributes.Strict.~this` is declared here
+fail_compilation\dtorfields_attributes.d(117):        it's @system because of the following field's destructors:
+fail_compilation\dtorfields_attributes.d(117):         - member
+fail_compilation\dtorfields_attributes.d(117): Error: `@nogc` constructor `dtorfields_attributes.Strict.this` cannot call non-@nogc destructor `dtorfields_attributes.Strict.~this`
+fail_compilation\dtorfields_attributes.d(117):        it's non-@nogc because of the following field's destructors:
+fail_compilation\dtorfields_attributes.d(117):         - member
+---
+
+TODO: nothrow is not checked, but that seems to be an unrelated bug
```
Comment (wilzbach) [inline test/fail_compilation/dtorfields_attributes.d:19]: Better open a bugzilla issue and link it here (otherwise it gets forgotten)

File: src/dmd/expression.d [hunk 4]
```diff
@@ -1107,11 +1107,81 @@ extern (C++) abstract class Expression : ASTNode
...
+            if (dtor.generated)
+                continue;
+
+            if (!check(dtor)) // doesn't match check (e.g. is impure as well)
+                break;
+
+            // Sanity check
+            assert(!check(cast(DtorDeclaration) ad.fieldDtor));
+
+            errorSupplemental("it's %s because of the following field's destructors:", what);
```
Comment (wilzbach) [inline src/dmd/expression.d:1153]:
```suggestion
            errorSupplemental("it's `%.*s` because of the following field's destructors:", checkName.length, checkName.ptr);
```

File: src/dmd/expression.d [hunk 5]
```diff
@@ -1107,11 +1107,81 @@ extern (C++) abstract class Expression : ASTNode
...
+     * missmatch.
+     *
+     * Params:
+     *   sc = scope
+     *   f  = potential `DtorDeclaration`
+     *   check = current check (e.g. whether it's pure)
+     *   what = the kind of check (e.g. `"pure"`)
+     */
+    extern (D) void checkOverridenDtor(Scope* sc, FuncDeclaration f,
+                scope bool function(DtorDeclaration) check, const char* what
```
Comment (wilzbach) [inline src/dmd/expression.d:1131]:
```suggestion
                scope bool function(DtorDeclaration) check, string checkName
```

(new code should prefer to use D strings)

**Review by wilzbach — APPROVED** [review]
This will be a great help to the users trying to decipher the error messages!

I wonder whether it would make sense (for an extension/future work) to generalize this as deciphering a `@system` chain can often be hard if templates are involved.

**Comment by MoonlightSentinel** [comment]
> I wonder whether it would make sense (for an extension/future work) to generalize this as deciphering a `@system` chain can often be hard if templates are involved.

You mean something like this:

```
import core.stdc.stdio : puts;

void foo()()
{
     puts("Hello, World!");
}

void main() @safe
{
    foo!();
}
```

```
onlineapp.d(10): Error: @safe function D main cannot call @system function onlineapp.foo!().foo
onlineapp.d(3):        onlineapp.foo!().foo is declared here
onlineapp.d(3):        onlineapp.foo!().foo is inferred @system because of the following operations:
onlineapp.d(5):        -  puts("Hello, World!");
```

**Comment by MoonlightSentinel** [comment]
Added tracking for destructors s.t. you see the desctructor of a member of a member .... which causes the less-restrictive attribute inference.

**Comment by nordlow** [comment]
This is awesome diagnostics. How hard would it be to do the same to trace transitive mismatches for qualifiers `@safe`, `pure`, `nothrow`, `@nogc`?

**Comment by MoonlightSentinel** [comment]
This diagnostic was rather easy because we know exactly what t look for. A general check should be possible using a visitor but are propbably a lot harder to get right.
